### DISCUSSION TITLE
What should the languages.toml configuration file look like for Rust development?

### DISCUSSION THREAD

**Discussion by ArikRahman** [helix-editor/helix#3443] [Q&A] [answered]
Newbie here, I have no idea where to start on how to start programming with Rust exclusively in Helix. When I cargo run a file, there is no debugging or auto-completion or anything. The only confirmation I have of the config file working is the auto-pairs. I'm not sure rust-analyzer is doing anything because nothing noticeable happened when I started programming with it. I installed the rust-analyzer Language Server Binary  per instructions from their website so I'm sure everything is done correctly. My languages.toml config file looks like this:

[language.auto-pairs]
'(' = ')'
'{' = '}'
'[' = ']'
'"' = '"'
'`' = '`'

[[language]]
name = "rust"
scope = "source.rust"
injection-regex = "rust"
file-types = ["rs"]
roots = []
auto-format = true
comment-token = "//"
language-server = { command = "rust-analyzer" }
indent = { tab-width = 4, unit = "    " }

[language.config]
cargo = { loadOutDirsFromCheck = true }
procMacro = { enable = false }
diagnostics = { disabled = ["unresolved-proc-macro"] }[[grammar]]
name = "rust"
source = { git = "https://github.com/tree-sitter/tree-sitter-rust", rev = "a360da0a29a19c281d08295a35ecd0544d2da211" }

[[grammar]]
name = "rust"
source = { git = "https://github.com/tree-sitter/tree-sitter-rust", rev = "a360da0a29a19c281d08295a35ecd0544d2da211" }

What am I missing or doing wrong?

**Comment by sbromberger** [comment, marked as answer]
Here's my `languages.toml` that's in my main configuration directory. It seems to work well for Rust. YMMV. You shouldn't have to edit anything in the `runtime` directory.

```
[[language]]
name="rust"
[language.config]
checkOnSave = {command = "clippy"}
cargo = {allFeatures = true}
procMacro = {enable = true }
auto-format = true
```

**Reply by cor** [reply]
This works very well. Thanks for sharing!

**Reply by David-Else** [reply]
I am not sure if it is working, I am 100% new to rust. The language server is running, but should this give on screen errors? It does get compiler error `cannot find value `x` in this scope`:
```
fn main() {
    x = 5;
    println!("x has the value {}", x);
}
```

I have

```
[[language]]
name="rust"
[language.config]
checkOnSave = {command = "clippy"}
cargo = {allFeatures = true}
```
How can I test if clippy is working too? Thanks!

(`auto-format = true` is inherited and not needed, works anyway)

**Comment by ArikRahman** [comment]
Thank you Mr. Bromberger. This unfortunately did not change much for me but I'll stick with Helix until I can understand the documentation better!

**Reply by sbromberger** [reply]
Next troubleshooting step: make sure your `rust-analyzer` is executable (on your `$PATH`).

**Reply by ArikRahman** [reply]
OMG! That worked! In addition I had to type in the command ":set-language Rust". Rookie mistake from me. I'm just glad I figured it out so I can join y'all in the wonders of modal editing!

**Reply by sbromberger** [reply]
> In addition I had to type in the command ":set-language Rust".

You shouldn't have to do this if you've created a Rust package using `cargo init` and are editing `.rs` files inside of `src/`.

**Reply by ArikRahman** [reply]
> > In addition I had to type in the command ":set-language Rust".
> 
> You shouldn't have to do this if you've created a Rust package using `cargo init` and are editing `.rs` files inside of `src/`.

Oh that makes sense. I used cargo new to initialize the project and I did that outside of Helix. I will keep that in mind for next time.

**Reply by sbromberger** [reply]
(@ArikRahman - suggest you mark this solved/closed.)

**Reply by Omnikar** [reply]
> > > In addition I had to type in the command ":set-language Rust".
> > 
> > 
> > You shouldn't have to do this if you've created a Rust package using `cargo init` and are editing `.rs` files inside of `src/`.
> 
> Oh that makes sense. I used cargo new to initialize the project and I did that outside of Helix. I will keep that in mind for next time.

`cargo init` for an existing directory and `cargo new` for a new directory both work the same.
